content/flux/v0.x/data-types/dynamic.md

```diff
@@ -0,0 +1,622 @@
+to be the same type. JSON arrays allow elements to be different types.
+
+With Flux records (the corollary of a JSON object), the properties in the record
+are part of the record's type.
+If you have a JSON array of objects where the objects' schemas are different,
+the array cannot be parsed as a Flux array because the types of the records in
+the array do not match.
+
+{{< expand-wrapper >}}
+{{% expand "View examples of JSON arrays that cannot be directly parsed into Flux arrays" %}}
+
+#### Array of scalar values
+
+The following array is a valid JSON array, but an invalid Flux array because
+each element in the array is a different type.
+
+```json
+["three", 2, true]
+```
+
+#### Array of objects
+
+The following array is a valid JSON array, but an invalid Flux array.
+The schemas of the objects in the array do not match, so when converted to Flux
+records, their types do not match and the array is invalid.
+
+```json
+[
+    {"first-name": "John", "last-name": "Doe", "age": 42},
+    {"first-name": "Jane", "last-name": "Doe"},
+    {"first-name": "James", "last-name": "Doe", "age": 15}
+]
+```
+
+{{% /expand %}}
+{{< /expand-wrapper >}}
+
+The dynamic type provides a way to resolve the differences between JSON and Flux types.
+
+## Dynamic type syntax
+
+Flux does not provide a literal syntax for dynamic types.
+To cast a value to a dynamic type:
+
+1.  Import the [`experimental/dynamic` package](/flux/v0.x/stdlib/experimental/dynamic/).
+2.  Use [`dynamic.dynamic()`](/flux/v0.x/stdlib/experimental/dynamic/dynamic/) to
+    convert a value to a dynamic type.
+
+```js
+import "experimental/dynamic"
+
+dynamic.dynamic(v: "Example string")
+
+// Returns dynamic(Example string)
+```
+
+## Reference values in a dynamic type
+
+When working dynamic records and arrays, use bracket and dot notation to reference
+values in each.
+
+##### Reference values in a dynamic record
+
+```js
+import "experimental/dynamic"
+
+record = {one: 1, two: 2, three: 3}
+dynamicRecord = dynamic.dynamic(v: record)
+
+dynamicRecord.one
+// Returns dynamic(1)
+
+dynamicRecord["two"]
+// Returns dynamic(2)
+```
+
+##### Reference values in a dynamic array
+
+```js
+import "experimental/dynamic"
+
+arr = ["a", "b", "c"]
+dynamicArr = dynamic.asArray(v: dynamic.dynamic(v: arr))
+
+dynamicArr[0]
+// Returns dynamic(a)
+```
+
+## Operate on dynamic types
+
+- [Convert dynamic types to Flux types](#convert-dynamic-types-to-flux-types)
+  - [Basic types](#basic-types)
+  - [Composite types](#composite-types)
+- [Include dynamic types in a table](#include-dynamic-types-in-a-table)
+- [Convert a JSON array to a Flux table](#convert-a-json-array-to-a-flux-table)
+- [Encode a dynamic type as JSON](#encode-a-dynamic-type-as-json)
+  - [Encode a dynamic record as JSON](#encode-a-dynamic-record-as-json)
+  - [Encode a dynamic array of different basic types as JSON](#encode-a-dynamic-array-of-different-basic-types-as-json)
+
+### Convert dynamic types to Flux types
```

Comment:
Something that _might be worth mentioning_ is the boundary at the "table" level (ie when trying to use dynamic values to write back to a bucket etc).

We have this requirement that dynamic values are cast to proper basics before they can be used "as column data."

I don't know the best approach, but I'd think saying something about this boundary (which is real and intentional). Using these cast functions is required to cross the threshold, but it may not be clear what the threshold is unless we state it.

If we introduce this concept, it might even be good to mention up top in the overview then elaborate/restate down here.

Reply:
:rofl: I now see (moments later) you've got a separate section about building flux tables, which could also be a place to mention _the boundary_.

I still think there might be room for my initial feedback about discussing _the boundary_ explicitly, and in multiple locations in the docs. You never know where a user will enter the conversation.